File: frontend/src/components/task/TaskDependencyService.js
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8082/api/tasks';

class TaskDependencyService {
    // Add a dependency between two tasks
    static async addDependency(prerequisiteTaskId, dependentTaskId) {
        try {
            const response = await axios.post(`${API_BASE_URL}/dependencies`, {
                prerequisiteTaskId,
                dependentTaskId
            });
            return response.data;
        } catch (error) {
            throw new Error(error.response?.data?.error || 'Failed to add dependency');
        }
    }

    // Get the dependency graph
    static async getDependencyGraph() {
        try {
            const response = await axios.get(`${API_BASE_URL}/dependencies/graph`);
            return response.data;
        } catch (error) {
            throw new Error('Failed to fetch dependency graph');
        }
    }

    // Get the critical path
    static async getCriticalPath() {
        try {
            const response = await axios.get(`${API_BASE_URL}/dependencies/critical-path`);
            return response.data;
        } catch (error) {
            throw new Error('Failed to fetch critical path');
        }
    }

    // Get impact analysis for a task
    static async getImpactAnalysis(taskId) {
        try {
            const response = await axios.get(`${API_BASE_URL}/dependencies/${taskId}/impact-analysis`);
            return response.data;
        } catch (error) {
            throw new Error('Failed to fetch impact analysis');
        }
    }

    // Remove a dependency
    static async removeDependency(prerequisiteTaskId, dependentTaskId) {
        try {
            const response = await axios.delete(`${API_BASE_URL}/dependencies/${prerequisiteTaskId}/${dependentTaskId}`);
            return response.data;
        } catch (error) {
            throw new Error('Failed to remove dependency');
        }
    }
}

export default TaskDependencyService;

File: dsa-project-backend/src/main/java/com/taskmate/dsaprojectbackend/task/TaskDependencyService.java
package com.taskmate.dsaprojectbackend.task;

import com.taskmate.dsaprojectbackend.common.TaskDependencyGraph;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class TaskDependencyService {
    
    @Autowired
    private TaskService taskService;
    
    private TaskDependencyGraph dependencyGraph;
    
    public TaskDependencyService() {
        this.dependencyGraph = new TaskDependencyGraph();
    }
    
    // Validate and add dependency, checking for cycles
    public boolean validateAndAddDependency(int prereqId, int dependentId) throws Exception {
        // Add the edge temporarily
        dependencyGraph.addEdge(prereqId, dependentId);
        
        // Check for cycles
        if (dependencyGraph.detectCycle()) {
            // Remove the edge if it creates a cycle
            dependencyGraph.removeEdge(prereqId, dependentId);
            throw new Exception("Circular dependency detected");
        }
        
        return true;
    }
    
    // Get task order (topological sort)
    public List<Integer> getTaskOrder() {
        return dependencyGraph.topologicalSort();
    }
    
    // Get critical path
    public List<Integer> getCriticalPath() {
        return dependencyGraph.findCriticalPath();
    }
    
    // Get all tasks affected by a delay in a specific task
    public List<Integer> getAffectedTasks(int taskId) {
        return dependencyGraph.getImpactAnalysis(taskId);
    }
    
    // Remove a dependency
    public void removeDependency(int prereqId, int dependentId) {
        dependencyGraph.removeEdge(prereqId, dependentId);
    }
    
    // Get the dependency graph
    public TaskDependencyGraph getDependencyGraph() {
        return dependencyGraph;
    }
}